File: ios/BleNative/Handlers/BufferHandler.swift
import Foundation

class BufferHandler {
    private var bufferToId: [Int32: Buffer] = [:]
    
    private var buffersToType: [BufferType: [Buffer]] = [:]
    
    private var currentBufferId: Int32 = 0
    
    private let deviceScanBufferId = Int32.max
    
    private let restoreStateBufferId = Int32.max - 1
    
    func buffer(forId id: Int32) -> Buffer? {
        return bufferToId[id]
    }
    
    func actionOnBuffer(_ buffer: Buffer, options: [String: Any], fulfillRequest: Bool = true) -> [Any]? {
        let strategy = options.bufferStrategy ?? .take
        let placement = options.bufferPlacement ?? .latest
        let chunkSize = options.bufferChunkSize
        let reversed = placement == .oldest
        
        switch strategy {
        case .take:
            return buffer.take(amount: chunkSize, reversed: reversed, fulfill: fulfillRequest)
        case .peek:
            return buffer.peek(amount: chunkSize, reversed: reversed, fulfill: fulfillRequest)
        }
    }
    
    func removeBuffer(_ buffer: Buffer) {
        bufferToId.removeValue(forKey: buffer.id)
        if var buffers = buffersToType[buffer.type], let index = buffers.index(where: { $0.id == buffer.id }) {
            buffers.remove(at: index)
            if buffers.count == 0 {
                buffersToType.removeValue(forKey: buffer.type)
            } else {
                buffersToType[buffer.type] = buffers
            }
        }  
    }
    
    func addBuffer(type: BufferType, relatedIdentifier: Int32? = nil) -> Buffer {
        let bufferId = nextBufferId(forType: type)
        let buffer = Buffer(id: bufferId, type: type, relatedIdentifier: relatedIdentifier)
        
        bufferToId[bufferId] = buffer
        var buffers: [Buffer] = buffersToType[type] ?? []
        buffers.append(buffer)
        buffersToType[type] = buffers
        return buffer
    }
    
    func appendBufferElement(_ element: Any, forType type: BufferType, relatedIdentifier: Int32? = nil) -> [Buffer] {
        guard let buffers = buffersToType[type]?.filter({ $0.relatedIdentifier == relatedIdentifier && $0.invalidated == false }) else { return [] }
        
        for buffer in buffers {
            buffer.append(element)
        }
        return buffers
    }
    
    func markBuffersInvalidated(reason: BleError, exceptTypes: [BufferType]) -> [Buffer] {
        return markBuffersInvalidated(
            reason: reason,
            bufferFilter: { $0.invalidated == false },
            exceptTypes: exceptTypes
        )
    }
    
    func markBuffersInvalidated(reason: BleError, relatedIdentifier relatedIdentifier: Int32, exceptTypes: [BufferType]) -> [Buffer] {
        return markBuffersInvalidated(
            reason: reason,
            bufferFilter: { $0.relatedIdentifier == relatedIdentifier && $0.invalidated == false },
            exceptTypes: exceptTypes
        )
    }
    
    private func markBuffersInvalidated(reason: BleError, bufferFilter: (Buffer) -> Bool, exceptTypes: [BufferType]) -> [Buffer] {
        let buffers = buffersToType
            .filter { !exceptTypes.contains($0.key) }
            .values
            .reduce([]) { (result: [Buffer], buffers: [Buffer]) -> [Buffer] in
                return result + buffers.filter(bufferFilter)    
        }
        for buffer in buffers {
            buffer.invalidate(reason: reason)
        }
        return buffers
    }
    
    func hasBuffer(withType type: BufferType, relatedIdentifier: Int32? = nil) -> Bool {
        if let buffers = buffersToType[type], buffers.contains(where: { $0.relatedIdentifier == relatedIdentifier }) {
            return true
        }
        return false
    }
    
    private func nextBufferId(forType type: BufferType) -> Int32 {
        switch type {
        case .stateRestore:
            return restoreStateBufferId
        case .scan:
            return deviceScanBufferId
        default:
            currentBufferId += 1
            return currentBufferId
        }
    }
}

File: android/src/main/java/com/polidea/blenative/handlers/BufferHandler.kt
package com.polidea.blenative.handlers

import com.polidea.blenative.BufferActionPlacement
import com.polidea.blenative.BufferActionStrategy
import com.polidea.blenative.models.BleError
import com.polidea.blenative.models.Buffer
import com.polidea.blenative.models.BufferType
import com.polidea.blenative.utils.bufferChunkSize
import com.polidea.blenative.utils.bufferPlacement
import com.polidea.blenative.utils.bufferStrategy

class BufferHandler {
    private var idToBuffer: MutableMap<Int, Buffer> = mutableMapOf()

    private var bufferTypeToBuffers: MutableMap<BufferType, MutableList<Buffer>> = mutableMapOf()

    private var currentBufferId = 0

    fun buffer(id: Int): Buffer? = idToBuffer[id]

    fun actionOnBuffer(buffer: Buffer, options: Map<String, Any>, fulfillRequest: Boolean = true): List<Any>? {
        val strategy = options.bufferStrategy ?: BufferActionStrategy.TAKE
        val placement = options.bufferPlacement ?: BufferActionPlacement.LATEST
        val chunkSize = options.bufferChunkSize
        val lastItems = placement == BufferActionPlacement.LATEST

        return when (strategy) {
            BufferActionStrategy.TAKE -> buffer.take(chunkSize, lastItems, fulfillRequest)
            BufferActionStrategy.PEEK -> buffer.take(chunkSize, lastItems, fulfillRequest)
        }
    }

    fun removeBuffer(buffer: Buffer) {
        idToBuffer.remove(buffer.id)
        val buffers = bufferTypeToBuffers[buffer.type] ?: return
        buffers.removeAll { it.id == buffer.id }
        if (buffers.size == 0) {
            bufferTypeToBuffers.remove(buffer.type)
        }
    }

    fun addBuffer(type: BufferType, relatedIdentifier: Int? = null): Buffer {
        val bufferId = nextBufferId(type)
        val buffer = Buffer(bufferId, type, relatedIdentifier)
        idToBuffer[bufferId] = buffer
        val buffers = bufferTypeToBuffers[type] ?: mutableListOf()
        buffers.add(buffer)
        bufferTypeToBuffers[type] = buffers
        return buffer
    }

    fun appendBufferElements(elements: List<Any>, type: BufferType, relatedIdentifier: Int? = null): List<Buffer> {
        return appendBufferElement({ it.appendAll(elements) }, type, relatedIdentifier)
    }

    fun appendBufferElement(element: Any, type: BufferType, relatedIdentifier: Int? = null): List<Buffer> {
        return appendBufferElement({ it.append(element) }, type, relatedIdentifier)
    }

    private fun appendBufferElement(onAppend: (Buffer) -> Unit, type: BufferType, relatedIdentifier: Int? = null): List<Buffer> {
        val buffers = bufferTypeToBuffers[type]?.filter {
            it.relatedIdentifier == relatedIdentifier
                    && !it.invalidated
        } ?: return emptyList()
        buffers.forEach { onAppend(it) }
        return buffers
    }

    fun markBuffersInvalidated(reason: BleError, exceptTypes: List<BufferType>): List<Buffer> {
        return markBuffersInvalidated(reason, { !it.invalidated }, exceptTypes)
    }

    fun markBuffersInvalidated(reason: BleError, relatedIdentifier: Int, exceptTypes: List<BufferType>): List<Buffer> {
        return markBuffersInvalidated(reason,
                { it.relatedIdentifier == relatedIdentifier && !it.invalidated },
                exceptTypes)
    }

    fun markBuffersInvalidated(reason: BleError, type: BufferType): List<Buffer> {
        return markBuffersInvalidated(reason, { it.type == type }, listOf())
    }

    fun hasBuffer(type: BufferType, relatedIdentifier: Int? = null): Boolean {
        return bufferTypeToBuffers[type]?.find { it.relatedIdentifier == relatedIdentifier } != null
    }

    private fun markBuffersInvalidated(reason: BleError,
                                       bufferFilter: (Buffer) -> Boolean,
                                       exceptTypes: List<BufferType>): List<Buffer> {
        val buffers = bufferTypeToBuffers
                .filter { !exceptTypes.contains(it.key) }
                .values
                .fold(mutableListOf<Buffer>()) { result, buffers ->
                    return@fold (result + buffers.filter(bufferFilter)).toMutableList()
                }
        buffers.forEach { it.invalidate(reason) }
        return buffers
    }

    private fun nextBufferId(type: BufferType): Int {
        return when (type) {
            BufferType.SCAN -> deviceScanBufferId
            BufferType.STATE_RESTORE -> restoreStateBufferId
            else -> ++currentBufferId
        }
    }


    companion object {
        const val deviceScanBufferId = Int.MAX_VALUE
        const val restoreStateBufferId = Int.MAX_VALUE - 1
    }
}